### PR TITLE
[Task] Replacing typeloader with class resolver

### LINKED ISSUE #14482: [Task]: Use ClassResolver instead of Typeloader
### Task description

Documents should still support class_overrides as described in the documentation: https://pimcore.com/docs/pimcore/11.0/Development_Documentation/Extending_Pimcore/Overriding_Models.html

In the documents use the model factory again instead of calling the Typeloader directly.
https://github.com/pimcore/pimcore/blob/11.x/models/Document.php#L211

Implement a generic classresolver to use instead of the typeloader.

Update the documentation to make it more clear, that the type_definitions are primarily used for new document types.
To override existing classes, you should use the class_overrides.

### PR DESCRIPTION
**Pull request by mattamon**
## Changes in this pull request
Resolves https://github.com/pimcore/pimcore/issues/14482

## Additional info
Using a class resolver instead of using an implementation loader
Class resolver will return FQCN. This can be used in the model factory to check for overrides.

Also consider this change in https://github.com/pimcore/pimcore/pull/14407 and https://github.com/pimcore/pimcore/pull/14310 once this is merged in 11.x

### CODE REVIEW

File: lib/Resolver/ResolverInterface.php [hunk 1]
```diff
@@ -15,18 +14,12 @@
...
 {
     /**
-     * @param string $name
-     * @param array $params
-     *
-     * @return Document
+     * Returns FQCN of class
      */
-    public function build(string $name, array $params = []): Document;
+    public function resolve(string $class, string $type): ?string;
```
Comment (brusch) [inline lib/Resolver/ResolverInterface.php]:
```suggestion
    public function resolve(string $name, string $type): ?string;
```
Maybe this should stay `$name` because actually it's just an id-string, not a class, right?

File: lib/Resolver/ClassResolver.php [hunk 2]
```diff
@@ -15,20 +14,28 @@
...
  * @internal
  */
-final class PrefixLoader extends BasePrefixLoader
+class ClassResolver implements ResolverInterface
 {
-    /**
-     * {@inheritdoc}
-     */
-    protected function normalizeName(string $name): string
+    public const TYPE_DOCUMENTS = 'documents';
```
Comment (brusch) [inline lib/Resolver/ClassResolver.php]: I'm just wondering whether we should consider the type in the resolver or if we should have a resolver service for each type. 

So basically having a `pimcore.class.resolver.asset` and a `pimcore.class.resolver.document` where we configure the map as a parameter (that was mapped in the extension). 

So basically, like this: 
```yaml
    pimcore.class.resolver.document:
        arguments:
            $map: '%pimcore.documents.type_definitions.map%'
```

**Review by brusch — CHANGES_REQUESTED** [review]

File: lib/Resolver/ResolverInterface.php (hunk 1 above)
Comment (mattamon) [reply lib/Resolver/ResolverInterface.php]: This is true, it could be a class, but in general, it is better to call it name.

File: lib/Resolver/ClassResolver.php (hunk 2 above)
Comment (mattamon) [reply lib/Resolver/ClassResolver.php]: This would be a more generic approach I agree. The main difference is that you have to get the correct resolver for the type in the code instead of adding the type as a param. 

I am just wondering if it would make sense to have `n` (in our case most likely 2 or 3) services instead of one?

Comment (brusch) [reply lib/Resolver/ClassResolver.php]: I'd prefer `n` services to be honest, it's just more generic and makes it easier to e.g. override/decorate/extend with custom resolvers if necessary.

Comment (mattamon) [reply lib/Resolver/ClassResolver.php]: Agreed!

File: lib/Model/AbstractModel.php [hunk 3]
```diff
@@ -282,6 +283,14 @@ protected static function getModelFactory()
         return \Pimcore::getContainer()->get('pimcore.model.factory');
     }
 
+    /**
+     * @return ClassResolver
+     */
+    protected static function getDocumentClassResolver(): ClassResolver
```
Comment (mattamon) [inline lib/Model/AbstractModel.php]: @brusch 
Should we use a this for every type or should rather use the container call directly where we need it, or like add a param, which resolver to get?

Comment (brusch) [reply lib/Model/AbstractModel.php]: I'd go on the container directly

**Comment by martineiber** [comment]
https://github.com/pimcore/pimcore/blob/0bafc434239a001c072a01623ba6bad27a139056/bundles/CoreBundle/Resources/config/pimcore/default.yaml#L211

We could remove the prefix section, since we do not have a PrefixLoader anymore. Right?

**Comment by mattamon** [comment]
> https://github.com/pimcore/pimcore/blob/0bafc434239a001c072a01623ba6bad27a139056/bundles/CoreBundle/Resources/config/pimcore/default.yaml#L211
> 
> We could remove the prefix section, since we do not have a PrefixLoader anymore. Right?

Yes we could remove it, but we also have to change the Configuration, because we use the addImplementationLoaderNode atm.

File: doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md [hunk 4]
```diff
@@ -2,13 +2,11 @@
 
 Previously, the only way to define custom documents was to create them in a special namespace `Pimcore\Model\Document`. This
```
Comment (kingjia90) [inline doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md]: IMHO this doc should be more in absolute terms, the relative "Previously" may lose the intended sense and purpose in the long run.
 
```suggestion
Prior 10.6, the only way to define custom documents was to create them in a special namespace `Pimcore\Model\Document`. This
```

Maybe the first line should be something that explain the current state first eg.
```
Defining custom documents can be done in the config via a static mapping from document type to class name.
```
and potentially should move the "historical description" behind or into the Upgrade Notes instead.

File: doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md [hunk 5]
```diff
@@ -41,3 +39,9 @@ pimcore:
             map:
                 book: \App\Model\Document\Book
 ```
+
+### Do not use to override a class
+
+The type_definitions should only be used to add new documents.
+If you want to override an existing class please use the `pimcore:models:class_overrides` instead.
+For a more detailed explanation see [Overriding Models](https://github.com/pimcore/pimcore/tree/11.x/doc/Development_Documentation/20_Extending_Pimcore/03_Overriding_Models.md).
```
Comment (kingjia90) [inline doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md]:
```suggestion
For a more detailed explanation see [Overriding Models](../../03_Overriding_Models.md).
```
relative path would be better, ~please also verify why the 11.x doc is missing, potentially a remote tracking merge problem from 10.6 into 11~ the path was wrong, there's an extra slug `Development_Documentation`, should be https://github.com/pimcore/pimcore/blob/11.x/doc/20_Extending_Pimcore/03_Overriding_Models.md instead

File: lib/Resolver/ClassResolver.php [hunk 6]
```diff
@@ -15,20 +14,22 @@
  *  @license    http://www.pimcore.org/license     GPLv3 and PCL
  */
 
-namespace Pimcore\Model\Document\TypeDefinition\Loader;
 
-use Pimcore\Loader\ImplementationLoader\PrefixLoader as BasePrefixLoader;
+namespace Pimcore\Resolver;
```
Comment (kingjia90) [inline lib/Resolver/ClassResolver.php:18]: i see that the file was renamed and would take the history of the PrefixLoader along, not sure if is worth to change this and avoid that, as these files are not related to each other

image

File: bundles/CoreBundle/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -999,7 +999,7 @@ private function addDocumentsNode(ArrayNodeDefinition $rootNode)
                 ->end()
             ->end();
 
-        $this->addImplementationLoaderNode($documentsNode, 'type_definitions');
+        $this->addClassResolverNode($documentsNode, 'type_definitions');
```
Comment (kingjia90) [inline bundles/CoreBundle/DependencyInjection/Configuration.php:1002]: Not sure if this would be a BC break
https://github.com/pimcore/pimcore/pull/14509/commits/bebf533a764932c74dd39a5f8d67f5a4b87a1f62
if one had previously set some custom prefixes, it would now throw
`Unrecognized option "prefixes" under "pimcore.documents.type_definitions`

**Review by kingjia90 — CHANGES_REQUESTED** [review]
Overall LGTM, didn't find any other issues

File: lib/Resolver/ClassResolver.php (hunk 6 above)
Comment (mattamon) [reply lib/Resolver/ClassResolver.php:18]: I tried to resolve this serveral times, because I noticed this too, but it would always take the history, even tough I created the file from scratch. :/

File: bundles/CoreBundle/DependencyInjection/Configuration.php (hunk 7 above)
Comment (mattamon) [reply bundles/CoreBundle/DependencyInjection/Configuration.php:1002]: This was introduced in 10.6 and never finalized so I would not consider this a BC.

Comment (kingjia90) [reply bundles/CoreBundle/DependencyInjection/Configuration.php:1002]: ah, right 👍🏻

File: lib/Resolver/ClassResolver.php (hunk 6 above)
Comment (kingjia90) [reply lib/Resolver/ClassResolver.php:18]: Ah ok, then nevermind, if you already tried, doesn't seem worth to spend more time into it.
Probably will auto-solve once squashed before merging into the main branch

File: doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md (hunk 4 above)
Comment (mattamon) [reply doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md]: I would leave it in the document to make sure that defining it in the namespace is still possible but reorder it

File: doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md (hunk 5 above)
Comment (mattamon) [reply doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md]: https://github.com/pimcore/pimcore/blob/11.x/doc/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md I don't see a problem with the 11.x?

Comment (kingjia90) [reply doc/Development_Documentation/20_Extending_Pimcore/13_Bundle_Developers_Guide/08_Adding_Document_Types.md]: - that one is the right one, the current url in the branch and before this PR is 
https://github.com/pimcore/pimcore/tree/11.x/doc/Development_Documentation/20_Extending_Pimcore/03_Overriding_Models.md

- I noticed that i've suggested wrong, it seems there's an extra `../`

**Review by kingjia90 — APPROVED** [review]
